pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_numa_aware.go

```diff
@@ -74,79 +69,12 @@ func (p *PolicyNUMAAware) Update() (err error) {
 		}
 	}()
 
-	var (
-		errList           []error
-		reclaimableMemory float64
-		data              metric.MetricData
-	)
-
-	availNUMAs := p.metaServer.CPUDetails.NUMANodes()
-
-	reclaimedCoresContainers := make([]*types.ContainerInfo, 0)
-	p.metaReader.RangeContainer(func(podUID string, containerName string, containerInfo *types.ContainerInfo) bool {
-		if p.reclaimedContainersFilter(containerInfo) {
-			reclaimedCoresContainers = append(reclaimedCoresContainers, containerInfo)
-			return true
-		}
-
-		nodeReclaim := p.conf.GetDynamicConfiguration().EnableReclaim
-		reclaimEnable, err := helper.PodEnableReclaim(context.Background(), p.metaServer, podUID, nodeReclaim)
-		if err != nil {
-			errList = append(errList, err)
-			return true
-		}
-
-		if containerInfo.IsNumaExclusive() && !reclaimEnable {
-			memset := machine.GetCPUAssignmentNUMAs(containerInfo.TopologyAwareAssignments)
-			if memset.IsEmpty() {
-				errList = append(errList, fmt.Errorf("container(%v/%v) TopologyAwareAssignments is empty", containerInfo.PodName, containerName))
-				return true
-			}
-			availNUMAs = availNUMAs.Difference(memset)
-		}
-		return true
-	})
-
-	err = errors.NewAggregate(errList)
-	if err != nil {
-		return err
-	}
-
-	availNUMATotal := float64(0)
-	for _, numaID := range availNUMAs.ToSliceInt() {
-		data, err = p.metaServer.GetNumaMetric(numaID, consts.MetricMemFreeNuma)
-		if err != nil {
-			return err
-		}
-		reclaimableMemory += data.Value
-		general.InfoS("reclaimable numa memory free", "numaID", numaID, "numaFree", general.FormatMemoryQuantity(data.Value))
-
-		data, err = p.metaServer.GetNumaMetric(numaID, consts.MetricMemTotalNuma)
-		if err != nil {
-			return err
-		}
-		availNUMATotal += data.Value
-		general.InfoS("reclaimable numa memory total", "numaID", numaID, "numaTotal", general.FormatMemoryQuantity(data.Value))
-
-	}
-
-	for _, container := range reclaimedCoresContainers {
-		reclaimableMemory += container.MemoryRequest
-	}
-
-	watermarkScaleFactor, err := p.metaServer.GetNodeMetric(consts.MetricMemScaleFactorSystem)
+	h, err := helper.CalcPolicyNUMAAwareHeadroom(p.conf, p.metaReader, p.metaServer, p.essentials)
```

Comment:
There's no need to abstract all the headroom calculation logic. You could simply create a function to obtain availNUMAs and reclaimedCoresContainers. Both the provision and headroom policy can utilize this function, while other logic can remain independent.

Reply:
done